bash-cache.sh

```diff
@@ -278,31 +256,30 @@ bc::cache() {
     $_bc_enabled || { bc::orig::%func% "$@"; return; }
     ( bc::_cleanup & ) # Clean up stale caches in the background
 
-    local func="%func%" ttl="%ttl%" refresh="%refresh%" env=(%env%) args=("$@") cache_read_loc
+    local func="%func%" ttl="%ttl%" refresh="%refresh%" env=(%env%) args=("$@") exit cache_read_loc
     bc::_set_cache_read_loc
 
-    # Read from cache - capture output once to avoid races
-    # Note redirecting stderr to /dev/null comes first to suppress errors due to missing stdin
-    local out err exit
-    bc::_read_input out 2>/dev/null < "${cache_read_loc}/out" || true
-    bc::_read_input err 2>/dev/null < "${cache_read_loc}/err" || true
-    bc::_read_input exit 2>/dev/null < "${cache_read_loc}/exit" || true
-
-    if [[ -z "$exit" ]]; then
-      # No cache, execute in foreground
+    while true; do
+      { {
+        exit=$(< "${cache_read_loc}/exit")
+        # if exit is missing/empty we raced with a cleanup, disregard cache
+        if [[ -n "$exit" ]]; then
+          if (( refresh > 0 )) && ! bc::_newer_than "${cache_read_loc}/exit" "$refresh"; then
+            # Cache exists but is old, refresh in background
+            ( bc::_write_cache & )
+          fi
+          cat <&3 >&1 # stdout
```

Comment:
Do these only work if `cat` is a builtin vs. an executable? If so, might wanna note that in a comment. If they only work if it's an executable, `command cat` might be in order given that it's pretty important.

Reply:
What circumstances would you expect to be problematic? AFAIK `cat` is never a builtin, but introducing a wrapper function `cat() { command cat "$@"; }` doesn't appear to cause trouble. Nevertheless adding `command` here seems like a good idea, so I did that.

I had tried to find a way to avoid needing to call into `cat` at all, but I haven't found an approach that works yet. Any ideas?

Reply:
Sorry, that was imprecisely phrased. Yes, "command" was what I was after, in case it accidentally called into someone's poorly-customized alias/shell function.

As for a pure-bash version, this *might* work, but might rely on a separate shell being launched: http://rus.har.mn/blog/2009-05-28/pure-bash-cat/

Reply:
I'd actually came across that post, and left a comment because there's a bug in it ;) I'll stick with `command cat` for now.

Reply:
Funny story, that blog post is actually written by my teammate @eatnumber1 and I didn't even notice at the time 😄

Reply:
Two of you in the same room is not something the world is ready for.